src/java_bytecode/expr2java.h

```diff
@@ -50,4 +52,40 @@ class expr2javat:public expr2ct
 std::string expr2java(const exprt &expr, const namespacet &ns);
 std::string type2java(const typet &type, const namespacet &ns);
 
+/// Convert floating point number to a string without printing
+/// unnecessary zeros. Prints decimal if precision is not lost.
+/// Prints hexadecimal otherwise, and/or java-friendly NaN and Infinity
+template <typename float_type>
+std::string floating_point_to_java_string(float_type value)
+{
+  static_assert(sizeof(float) == 4, "Float should be 32-bit type");
+  static_assert(sizeof(double) == 8, "Double should be 64-bit type");
+  const auto is_float = std::is_same<float_type, float>::value;
+  static const std::string class_name = is_float ? "Float" : "Double";
+  if(std::isnan(value))
+    return class_name + ".NaN";
+  if(std::isinf(value))
+    return class_name + ".POSITIVE_INFINITY";
+  if(std::isinf(-value))
+    return class_name + ".NEGATIVE_INFINITY";
+  const auto raw_decimal = std::to_string(value);
+  const auto decimal = raw_decimal.find('.') == std::string::npos
+                         ? raw_decimal + ".0"
+                         : raw_decimal;
+  const bool is_lossless = std::stod(decimal == value);
+  const auto lossless = [&] { // NOLINT
+    if(is_lossless)
```

Comment:
This is backwards I think, we should use hexfloat if the decimal representation is **lossy**